File: web/backend/NodeJS-services/database/models/ScheduleFormat.js
const mongoose = require('mongoose');
const { COLLECTIONS } = require('../constants');

/**
 * ScheduleFormat Schema
 * Represents a time slot in the schedule
 */
const ScheduleFormatSchema = new mongoose.Schema({
  slotId: {
    type: Number,
    required: true,
    unique: true,
    auto: true
  },
  slotNumber: {
    type: Number,
    required: true
  },
  slotName: {
    type: String,
    required: true
  },
  dayOfWeek: {
    type: String,
    enum: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
    required: true
  },
  startTime: {
    type: String,
    required: true
  },
  endTime: {
    type: String,
    required: true
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true,
  versionKey: false,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
  collection: 'ScheduleFormat'
});

// Virtual for getting schedules using this slot
ScheduleFormatSchema.virtual('schedules', {
  ref: 'ClassSchedule',
  localField: 'slotId',
  foreignField: 'slotId',
  justOne: false
});

module.exports = mongoose.model('ScheduleFormat', ScheduleFormatSchema); 

File: web/backend/Python-services/src/models/ScheduleFormat.py
"""
ScheduleFormat model for FAMS
"""
from datetime import datetime
from typing import Optional, ClassVar, Dict, Any
from pydantic import BaseModel, Field
from ..constants import COLLECTIONS


class ScheduleFormat(BaseModel):
    """ScheduleFormat model representing time slots in the schedule"""
    slotId: int = Field(..., description="Unique identifier for the slot")
    slotNumber: int = Field(..., description="Number of the slot")
    dayOfWeek: str = Field(..., description="Day of week (Monday, Tuesday, etc.)")
    startTime: str = Field(..., description="Start time in HH:MM format")
    endTime: str = Field(..., description="End time in HH:MM format")
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    isActive: bool = Field(default=True)
    collection: str = COLLECTIONS['SCHEDULE_FORMAT']
    
    @classmethod
    def from_csv_row(cls, row: Dict[str, Any], slot_id: int) -> 'ScheduleFormat':
        """Create a ScheduleFormat instance from a CSV row"""
        return cls(
            slotId=slot_id,
            slotNumber=int(row.get("SlotNumber", 1)),
            dayOfWeek=row.get("DayOfWeek", "Monday"),
            startTime=row.get("StartTime", "08:00"),
            endTime=row.get("EndTime", "09:00"),
            isActive=True
        ) 